File: backend/app/api/phase3.py
from __future__ import annotations

import asyncio
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Literal
from uuid import UUID

from fastapi import APIRouter, Body, Header, Query, Request
from fastapi.responses import FileResponse, StreamingResponse
from pydantic import BaseModel, Field, ValidationError

from app.evidence.schemas import LiveTelemetryInput
from app.core.errors import DatabaseUnavailableError
from app.visualization.schemas import VisualizationResolveRequest, VisualizationResolvedV1


router = APIRouter(tags=["phase3"])


def service(request: Request):
    value = request.app.state.phase3_service
    if value is None:
        raise DatabaseUnavailableError()
    return value


class LiveSessionCreate(BaseModel):
    label: str = Field(default="Controlled live capture", min_length=1, max_length=200)
    source_ids: list[str] = Field(min_length=1, max_length=20)
    stale_after_seconds: int = Field(default=30, ge=5, le=300)


class AssistantSessionCreate(BaseModel):
    scope: Literal["auto", "all_cases", "specific_case", "selected_references"] = "auto"
    case_ids: list[int] = Field(default_factory=list, max_length=10)
    reference_ids: list[str] = Field(default_factory=list, max_length=50)


class AssistantMessageCreate(BaseModel):
    question: str = Field(min_length=1, max_length=4000)
    include_evidence: bool | None = None
    visualization_mode: Literal["none", "auto", "timeline", "severity_distribution", "event_activity", "entity_graph", "top_entities", "top_findings"] = "none"


class AssistantSessionPublic(BaseModel):
    session_id: UUID
    scope: Literal["auto", "all_cases", "specific_case", "selected_references"]
    case_ids: list[int]
    reference_ids: list[str]
    title: str
    message_count: int = Field(ge=0)
    created_at: datetime
    updated_at: datetime


class AssistantErrorPublic(BaseModel):
    code: str
    message: str
    retryable: bool


class AssistantJobPublic(BaseModel):
    job_id: UUID
    session_id: UUID
    status: Literal["queued", "processing", "completed", "failed"]
    error: AssistantErrorPublic | None
    created_at: datetime
    updated_at: datetime


class AssistantMessagePublic(BaseModel):
    message_id: UUID
    session_id: UUID
    role: Literal["user", "assistant"]
    kind: str
    text: str
    citations: list[str]
    caveats: list[str]
    visualization: dict | None
    model: str | None
    created_at: datetime


class AssistantSessionPage(BaseModel):
    items: list[AssistantSessionPublic]
    page: int
    page_size: int
    total: int


class AssistantMessagePage(BaseModel):
    items: list[AssistantMessagePublic]
    page: int
    page_size: int
    total: int


class ReportCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    sections: list[str] = Field(default_factory=lambda: ["executive_summary", "scope", "findings", "evidence", "timeline", "recommendations"])
    narrative: str | None = Field(default=None, max_length=10000)


class Approval(BaseModel):
    approver: str = Field(min_length=1, max_length=120)
    confirmed: Literal[True]


class EmailDraftBody(BaseModel):
    recipient: str = Field(pattern=r"^[^@\s]+@[^@\s]+\.[^@\s]+$", max_length=320)
    subject: str = Field(min_length=1, max_length=300)
    body: str = Field(min_length=1, max_length=20000)


@router.post("/cases/{case_id}/live-sessions", status_code=201)
def start_session(case_id: int, body: LiveSessionCreate, request: Request):
    return service(request).start_live_session(case_id, body.label, body.source_ids, body.stale_after_seconds, getattr(request.state, "request_id", None))


@router.get("/cases/{case_id}/live-sessions")
def list_sessions(case_id: int, request: Request):
    rows = service(request).db.execute("SELECT session_id FROM live_sessions WHERE case_id=? ORDER BY started_at DESC", (case_id,)).fetchall()
    return {"items": [service(request).get_live_session(row[0]) for row in rows], "page": 1, "page_size": len(rows), "total": len(rows)}


@router.get("/live-sessions/{session_id}")
def get_session(session_id: UUID, request: Request):
    return service(request).get_live_session(str(session_id))


@router.post("/live-sessions/{session_id}/stop")
def stop_session(session_id: UUID, request: Request):
    return service(request).stop_live_session(str(session_id), getattr(request.state, "request_id", None))


@router.post("/live/telemetry", status_code=202)
async def receive_telemetry(request: Request, x_traceveil_source_token: str = Header(alias="X-Traceveil-Source-Token")):
    raw = await request.body()
    svc = service(request)
    if len(raw) > svc.settings.live_max_payload_bytes:
        raise OverflowError("live_payload_too_large")
    request_id = getattr(request.state, "request_id", None)
    client_ip = request.client.host if request.client else None
    try:
        payload = json.loads(raw)
        telemetry = LiveTelemetryInput.model_validate(payload)
    except (json.JSONDecodeError, ValidationError) as exc:
        payload = payload if "payload" in locals() and isinstance(payload, dict) else {}
        source_id = payload.get("source_id")
        if not isinstance(source_id, str):
            # TV5-12: audit before raising so a rogue-source request
            # produces a bounded, structured record even when the body
            # does not carry a usable source_id.
            svc.record_auth_failure(None, "missing_source_id", request_id=request_id, client_ip=client_ip)
            raise PermissionError("invalid_live_source_token") from exc
        try:
            svc.verify_source(source_id, x_traceveil_source_token)
        except PermissionError:
            svc.record_auth_failure(source_id, "invalid_token", request_id=request_id, client_ip=client_ip)
            raise
        svc.record_malformed(payload.get("case_id"), source_id, "malformed_live_telemetry", str(exc), raw.decode("utf-8", errors="replace")[:65536])
        raise ValueError("malformed_live_telemetry") from exc

    try:
        return svc.submit_live(telemetry, x_traceveil_source_token, request_id)
    except PermissionError:
        # TV5-12: well-formed body but bad token still produces one
        # rate-limited audit row before re-raising.
        svc.record_auth_failure(telemetry.source_id, "invalid_token", request_id=request_id, client_ip=client_ip)
        raise


@router.get("/live/receipts/{receipt_id}")
def receipt(receipt_id: UUID, request: Request):
    return service(request).get_live_receipt(str(receipt_id))


@router.get("/cases/{case_id}/live/devices")
def live_devices(case_id: int, request: Request, session_id: UUID | None = None):
    items = service(request).list_devices(case_id, str(session_id) if session_id else None)
    return {"items": items, "page": 1, "page_size": len(items), "total": len(items)}


@router.get("/cases/{case_id}/live/metrics")
def live_metrics(case_id: int, request: Request, session_id: UUID | None = None):
    svc = service(request); session = str(session_id) if session_id else None
    where = "case_id=?" + (" AND session_id=?" if session else ""); values: tuple[Any, ...] = (case_id, session) if session else (case_id,)
    devices = svc.db.execute(f"SELECT COUNT(*),COALESCE(SUM(event_count),0) FROM device_states WHERE {where}", values).fetchone()
    malformed = svc.db.execute("SELECT COUNT(*) FROM live_ingest_issues WHERE case_id=?" + (" AND session_id=?" if session else ""), values).fetchone()[0]
    alerts = svc.db.execute("SELECT COUNT(*) FROM live_alert_first_seen WHERE case_id=?", (case_id,)).fetchone()[0]
    return {"case_id": case_id, "session_id": session, "device_count": devices[0], "event_count": devices[1], "alert_count": alerts, "malformed_count": malformed, "updated_at": svc.db.execute("SELECT MAX(last_seen_at) FROM device_states WHERE case_id=?", (case_id,)).fetchone()[0]}


@router.get("/live/stream")
async def live_stream(request: Request, case_id: int | None = None, session_id: UUID | None = None, topics: str = "", last_event_id: int | None = Header(default=None, alias="Last-Event-ID")):
    selected = [item.strip() for item in topics.split(",") if item.strip()]
    start = last_event_id or 0
    async def generate():
        cursor, last_heartbeat = start, 0.0
        while not await request.is_disconnected():
            messages = service(request).stream_after(cursor, case_id, str(session_id) if session_id else None, selected)
            if messages:
                for message in messages:
                    cursor = message["stream_id"]
                    envelope = {"schema_version": "1.0", "id": cursor, "topic": message["topic"], "case_id": message["case_id"], "session_id": message["session_id"], "occurred_at": message["occurred_at"], "payload": message["payload"]}
                    yield f"id: {cursor}\nevent: {message['topic']}\ndata: {json.dumps(envelope, separators=(',', ':'))}\n\n"
            elif asyncio.get_running_loop().time() - last_heartbeat >= 10:
                last_heartbeat = asyncio.get_running_loop().time()
                yield f"event: heartbeat\ndata: {json.dumps({'schema_version':'1.0','topic':'heartbeat'})}\n\n"
            await asyncio.sleep(0.25)
    return StreamingResponse(generate(), media_type="text/event-stream", headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no", "Connection": "keep-alive"})


@router.get("/cases/{case_id}/audit")
def audit(case_id: int, request: Request, page: int = Query(1, ge=1), page_size: int = Query(50, ge=1, le=200)):
    svc = service(request); total = svc.db.execute("SELECT COUNT(*) FROM audit_events WHERE case_id=?", (case_id,)).fetchone()[0]
    rows = svc.db.execute("SELECT * FROM audit_events WHERE case_id=? ORDER BY occurred_at DESC,audit_id DESC LIMIT ? OFFSET ?", (case_id, page_size, (page - 1) * page_size)).fetchall()
    items = []
    for row in rows:
        item = dict(row); item["details"] = json.loads(item.pop("details_json")); items.append(item)
    return {"items": items, "page": page, "page_size": page_size, "total": total}


@router.post("/assistant/sessions", status_code=201, response_model=AssistantSessionPublic)
def create_assistant_session(body: AssistantSessionCreate, request: Request):
    return service(request).create_assistant_session(body.scope, body.case_ids, body.reference_ids)


@router.get("/assistant/sessions", response_model=AssistantSessionPage)
def assistant_sessions(request: Request, case_id: int | None = None):
    items = service(request).list_assistant_sessions(case_id)
    return {"items": items, "page": 1, "page_size": len(items), "total": len(items)}


@router.get("/assistant/sessions/{session_id}", response_model=AssistantSessionPublic)
def assistant_session(session_id: UUID, request: Request):
    return service(request).get_assistant_session(str(session_id))


@router.get("/assistant/sessions/{session_id}/messages", response_model=AssistantMessagePage)
def assistant_messages(session_id: UUID, request: Request):
    items = service(request).assistant_messages(str(session_id))
    return {"items": items, "page": 1, "page_size": len(items), "total": len(items)}


@router.post("/assistant/sessions/{session_id}/messages", status_code=202, response_model=AssistantJobPublic)
def assistant_message(session_id: UUID, body: AssistantMessageCreate, request: Request):
    return service(request).submit_assistant_message(
        str(session_id), body.question, body.include_evidence, body.visualization_mode
    )


@router.get("/assistant/jobs/{job_id}", response_model=AssistantJobPublic)
def assistant_job(job_id: UUID, request: Request):
    return service(request).get_assistant_job(str(job_id))


@router.post("/visualizations/resolve", response_model=VisualizationResolvedV1)
def resolve_visualization(body: VisualizationResolveRequest, request: Request):
    """Resolve a validated layout exclusively from persisted investigation data."""
    return service(request).visualizations.resolve(body.layout)


@router.get("/cases/{case_id}/visualizations/fallback", response_model=VisualizationResolvedV1)
def fallback_visualization(
    case_id: int,
    request: Request,
    intent: Literal["overview", "timeline", "risk", "correlation", "alerts", "evidence", "live", "severity_distribution", "event_activity", "entity_graph", "top_entities", "top_findings"] = "overview",
    analysis_id: UUID | None = None,
):
    selector = str(analysis_id) if analysis_id else "latest"
    layout = service(request).visualizations.fallback(case_id, intent, selector)
    return service(request).visualizations.resolve(layout)


@router.post("/cases/{case_id}/reports", status_code=201)
def create_report(case_id: int, body: ReportCreate, request: Request):
    return service(request).create_report(case_id, body.title, body.sections, body.narrative)


@router.get("/cases/{case_id}/reports")
def list_reports(case_id: int, request: Request):
    rows = service(request).db.execute("SELECT report_id FROM reports WHERE case_id=? ORDER BY created_at DESC", (case_id,)).fetchall()
    items = [service(request).get_report(row[0]) for row in rows]
    return {"items": items, "page": 1, "page_size": len(items), "total": len(items)}


@router.get("/reports/{report_id}")
def get_report(report_id: UUID, request: Request):
    return service(request).get_report(str(report_id))


@router.post("/reports/{report_id}/generate")
def generate_report(report_id: UUID, request: Request):
    return service(request).generate_report(str(report_id))


@router.post("/reports/{report_id}/approve")
def approve_report(report_id: UUID, body: Approval, request: Request):
    return service(request).approve_report(str(report_id), body.approver)


@router.get("/reports/{report_id}/export")
def export_report(report_id: UUID, request: Request):
    report = service(request).get_report(str(report_id))
    if report["status"] not in {"generated", "approved"} or not report["file_path"]:
        raise ValueError("generated_report_required")
    return FileResponse(Path(report["file_path"]), media_type="application/pdf", filename=f"traceveil-{report_id}.pdf", headers={"ETag": report["content_hash"]})


@router.post("/reports/{report_id}/email-drafts", status_code=201)
def create_email_draft(report_id: UUID, body: EmailDraftBody, request: Request):
    return service(request).create_email_draft(str(report_id), str(body.recipient), body.subject, body.body)


@router.get("/email-drafts/{draft_id}")
def get_email_draft(draft_id: UUID, request: Request):
    return service(request).get_email_draft(str(draft_id))


@router.patch("/email-drafts/{draft_id}")
def patch_email_draft(draft_id: UUID, body: EmailDraftBody, request: Request):
    return service(request).update_email_draft(str(draft_id), str(body.recipient), body.subject, body.body)


@router.post("/email-drafts/{draft_id}/approve")
def approve_email_draft(draft_id: UUID, body: Approval, request: Request):
    return service(request).approve_email(str(draft_id), body.approver)


@router.post("/email-drafts/{draft_id}/send")
def send_email_draft(draft_id: UUID, request: Request):
    return service(request).send_email(str(draft_id), getattr(request.state, "request_id", None))

File: frontend/src/api/phase3.ts
import { apiClient } from "./batch";

export interface LiveSession { session_id: string; case_id: number; label: string; source_ids: string[]; status: "active" | "completed"; stale_after_seconds: number; accepted_count: number; malformed_count: number; started_at: string; stopped_at: string | null; updated_at: string }
export interface LiveMetrics { case_id: number; session_id: string | null; device_count: number; event_count: number; alert_count: number; malformed_count: number; updated_at: string | null }
export interface DeviceState { case_id: number; session_id: string; device_id: string; source_id: string; last_seen_at: string; last_observed_at: string; latest_metrics: Record<string, string | number | boolean | null>; event_count: number; stale: boolean }
export type StreamTopic = "event.accepted" | "alert.created" | "device.updated" | "metrics.updated" | "session.updated" | "heartbeat";
export interface StreamEnvelope { schema_version: "1.0"; id?: number; topic: StreamTopic; case_id?: number; session_id?: string | null; occurred_at?: string; payload?: Record<string, unknown> }
export type VisualizationType = "timeline" | "risk_breakdown" | "severity_distribution" | "event_activity" | "entity_graph" | "evidence_table" | "alert_list" | "top_entities" | "top_findings";
export interface VisualizationComponent { id: string; type: VisualizationType; title: string; data_ref: string; span: 1 | 2 | 3; height: "compact" | "standard" | "tall" }
export interface VisualizationLayout { schema_version: "1.0"; layout_id: string; title: string; components: VisualizationComponent[] }
export interface ResolvedVisualization { schema_version: "1.0"; layout: VisualizationLayout; datasets: Record<string, unknown>; analysis_ids: Record<string, string> }
export interface AssistantSession { session_id: string; scope: "auto" | "all_cases" | "specific_case" | "selected_references"; case_ids: number[]; reference_ids: string[]; title: string; message_count: number; created_at: string; updated_at: string }
export interface AssistantMessage { message_id: string; role: "user" | "assistant"; kind: string; text: string; citations: string[]; caveats: string[]; visualization: VisualizationLayout | null; model: string | null; created_at: string }
export interface AssistantJob { job_id: string; session_id: string; status: "queued" | "processing" | "completed" | "failed"; error: { code: string; message: string; retryable: boolean } | null }
export type AssistantVisualizationMode = "none" | "auto" | "timeline" | "severity_distribution" | "event_activity" | "entity_graph" | "top_entities" | "top_findings";
export interface ReportRecord { report_id: string; case_id: number; title: string; sections: string[]; status: "draft" | "generated" | "approved"; narrative: string | null; content_hash: string | null; approved_by: string | null; approved_at: string | null; created_at: string; updated_at: string }
export interface EmailDraft { draft_id: string; report_id: string; recipient: string; subject: string; body: string; status: "draft" | "approved" | "sent" | "delivery_unknown"; approved_at: string | null; delivery_error: string | null }
export interface AuditEvent { audit_id: string; case_id: number | null; action: string; actor: string; subject_type: string; subject_id: string | null; request_id: string | null; details: Record<string, unknown>; occurred_at: string }
interface Page<T> { items: T[]; total: number; page: number; page_size: number }

export const phase3Api = {
  startSession(caseId: number, sourceIds = ["live-lab-01"], signal?: AbortSignal) { return apiClient.request<LiveSession>(`/cases/${caseId}/live-sessions`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ label: "Controlled live capture", source_ids: sourceIds, stale_after_seconds: 30 }) }); },
  sessions(caseId: number, signal?: AbortSignal) { return apiClient.request<Page<LiveSession>>(`/cases/${caseId}/live-sessions`, { signal }); },
  stopSession(sessionId: string, signal?: AbortSignal) { return apiClient.request<LiveSession>(`/live-sessions/${sessionId}/stop`, { method: "POST", signal }); },
  metrics(caseId: number, sessionId?: string, signal?: AbortSignal) { return apiClient.request<LiveMetrics>(`/cases/${caseId}/live/metrics${sessionId ? `?session_id=${encodeURIComponent(sessionId)}` : ""}`, { signal }); },
  devices(caseId: number, sessionId?: string, signal?: AbortSignal) { return apiClient.request<Page<DeviceState>>(`/cases/${caseId}/live/devices${sessionId ? `?session_id=${encodeURIComponent(sessionId)}` : ""}`, { signal }); },
  createAssistantSession(caseIds: number[] = [], referenceIds: string[] = [], signal?: AbortSignal) { return apiClient.request<AssistantSession>("/assistant/sessions", { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ scope: referenceIds.length ? "selected_references" : caseIds.length ? "specific_case" : "auto", case_ids: caseIds, reference_ids: referenceIds }) }); },
  assistantSessions(caseId?: number | null, signal?: AbortSignal) { return apiClient.request<Page<AssistantSession>>(`/assistant/sessions${caseId ? `?case_id=${caseId}` : ""}`, { signal }); },
  assistantSession(sessionId: string, signal?: AbortSignal) { return apiClient.request<AssistantSession>(`/assistant/sessions/${sessionId}`, { signal }); },
  ask(sessionId: string, question: string, includeEvidence: boolean, visualizationMode: AssistantVisualizationMode, signal?: AbortSignal) { return apiClient.request<AssistantJob>(`/assistant/sessions/${sessionId}/messages`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ question, include_evidence: includeEvidence, visualization_mode: visualizationMode }) }); },
  assistantJob(jobId: string, signal?: AbortSignal) { return apiClient.request<AssistantJob>(`/assistant/jobs/${jobId}`, { signal }); },
  messages(sessionId: string, signal?: AbortSignal) { return apiClient.request<Page<AssistantMessage>>(`/assistant/sessions/${sessionId}/messages`, { signal }); },
  resolveVisualization(layout: VisualizationLayout, signal?: AbortSignal) { return apiClient.request<ResolvedVisualization>("/visualizations/resolve", { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ layout }) }); },
  fallbackVisualization(caseId: number, analysisId?: string, signal?: AbortSignal) { const query = new URLSearchParams({ intent: "overview" }); if (analysisId) query.set("analysis_id", analysisId); return apiClient.request<ResolvedVisualization>(`/cases/${caseId}/visualizations/fallback?${query}`, { signal }); },
  reports(caseId: number, signal?: AbortSignal) { return apiClient.request<Page<ReportRecord>>(`/cases/${caseId}/reports`, { signal }); },
  createReport(caseId: number, title: string, signal?: AbortSignal) { return apiClient.request<ReportRecord>(`/cases/${caseId}/reports`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ title }) }); },
  generateReport(reportId: string, signal?: AbortSignal) { return apiClient.request<ReportRecord>(`/reports/${reportId}/generate`, { method: "POST", signal }); },
  approveReport(reportId: string, approver: string, signal?: AbortSignal) { return apiClient.request<ReportRecord>(`/reports/${reportId}/approve`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ approver, confirmed: true }) }); },
  downloadReport(reportId: string, signal?: AbortSignal) { return apiClient.download(`/reports/${reportId}/export`, signal); },
  createEmailDraft(reportId: string, recipient: string, subject: string, body: string, signal?: AbortSignal) { return apiClient.request<EmailDraft>(`/reports/${reportId}/email-drafts`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ recipient, subject, body }) }); },
  approveEmail(draftId: string, approver: string, signal?: AbortSignal) { return apiClient.request<EmailDraft>(`/email-drafts/${draftId}/approve`, { method: "POST", signal, headers: { "Content-Type": "application/json" }, body: JSON.stringify({ approver, confirmed: true }) }); },
  sendEmail(draftId: string, signal?: AbortSignal) { return apiClient.request<EmailDraft>(`/email-drafts/${draftId}/send`, { method: "POST", signal }); },
  audit(caseId: number, signal?: AbortSignal) { return apiClient.request<Page<AuditEvent>>(`/cases/${caseId}/audit?page=1&page_size=100`, { signal }); },
};
